### PR TITLE
[Beam-13153] [Playground] Add validators and preparators for go code

### PR DESCRIPTION
**Pull request by pavel-avilov**
[[BEAM-13153]](https://issues.apache.org/jira/browse/BEAM-13153)

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

`ValidatesRunner` compliance status (on master branch)
--------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Lang</th>
      <th>ULR</th>
      <th>Dataflow</th>
      <th>Flink</th>
      <th>Samza</th>
      <th>Spark</th>
      <th>Twister2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Go</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          Build Status
        </a>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>Java</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_ULR/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Streaming/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2_Streaming/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
    <tr>
      <td>Python</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>XLang</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
  </tbody>
</table>

Examples testing status on various runners
--------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Lang</th>
      <th>ULR</th>
      <th>Dataflow</th>
      <th>Flink</th>
      <th>Samza</th>
      <th>Spark</th>
      <th>Twister2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Go</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>Java</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Java11_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_Examples_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>Python</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>XLang</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
  </tbody>
</table>

Post-Commit SDK/Transform Integration Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Go</th>
      <th>Java</th>
      <th>Python</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
  </tbody>
</table>

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>---</th>
      <th>Java</th>
      <th>Python</th>
      <th>Go</th>
      <th>Website</th>
      <th>Whitespace</th>
      <th>Typescript</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Non-portable</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
    <tr>
      <td>Portable</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_GoPortable_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
  </tbody>
</table>

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Review by ilya-kozyrev — APPROVED** [review]

File: playground/backend/internal/preparators/go_preparators.go [hunk 1]
```diff
@@ -15,7 +15,30 @@
...
+// formatCode formats go code
+func formatCode(args ...interface{}) error {
+	filePath := args[0].(string)
+	file := filepath.Base(filePath)
+	fileDir := filepath.Dir(filePath)
+	cmd := exec.Command("go", "fmt", file)
+	cmd.Dir = fileDir
+	stdout, err := cmd.CombinedOutput()
+	if err != nil {
+		return errors.New(string(stdout))
```
Comment (AydarZaynutdinov) [inline playground/backend/internal/preparators/go_preparators.go:41]: I think we also need to return an error. May be something
```
return fmt.Errorf("%s: %s", like err.Error(), string(stdout))
```

File: playground/backend/internal/preparators/go_preparators_test.go [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+func TestMain(m *testing.M) {
+	err := setupPreparedFiles()
+	if err != nil {
+		logger.Fatal(err)
+	}
+	defer teardown()
+	m.Run()
+}
+
+func setupPreparedFiles() error {
```
Comment (AydarZaynutdinov) [inline playground/backend/internal/preparators/go_preparators_test.go:40]: Please add a comment.

File: playground/backend/internal/preparators/go_preparators_test.go [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+	type args struct {
+		filePath string
+	}
+	tests := []struct {
+		name string
+		args args
+		want *[]Preparator
+	}{
+		{
+			name: "all success",
```
Comment (AydarZaynutdinov) [inline playground/backend/internal/preparators/go_preparators_test.go:90]: Change the name of the test case. And add a comment for this test case.

File: playground/backend/internal/preparators/go_preparators_test.go [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+	type args struct {
+		args []interface{}
+	}
+	tests := []struct {
+		name    string
+		args    args
+		wantErr bool
+	}{
+		{
+			name:    "all success",
```
Comment (AydarZaynutdinov) [inline playground/backend/internal/preparators/go_preparators_test.go:118]: ditto.

File: playground/backend/internal/preparators/go_preparators_test.go [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+			args:    args{preparedArgs1},
+			wantErr: false,
+		},
+		{
+			name:    "file with errors",
```
Comment (AydarZaynutdinov) [inline playground/backend/internal/preparators/go_preparators_test.go:123]: Add a comment for this test case.

**Review by AydarZaynutdinov — CHANGES_REQUESTED** [review]

File: playground/backend/internal/preparators/go_preparators.go (hunk 1 above)
Comment (pavel-avilov) [reply playground/backend/internal/preparators/go_preparators.go:41]: the stdout already contain this error

File: playground/backend/internal/preparators/go_preparators.go [hunk 6]
```diff
@@ -15,7 +15,30 @@
...
+	formatCodePreparator := Preparator{Prepare: formatCode, Args: preparatorArgs}
+	return &[]Preparator{formatCodePreparator}
+}
+
+// formatCode formats go code
+func formatCode(args ...interface{}) error {
+	filePath := args[0].(string)
+	file := filepath.Base(filePath)
+	fileDir := filepath.Dir(filePath)
+	cmd := exec.Command("go", "fmt", file)
```
Comment (daria-malkova) [inline playground/backend/internal/preparators/go_preparators.go:37]: constants

File: playground/backend/internal/preparators/go_preparators.go [hunk 7]
```diff
@@ -15,7 +15,30 @@
...
+	preparatorArgs := make([]interface{}, 1)
+	preparatorArgs[0] = filePath
+	formatCodePreparator := Preparator{Prepare: formatCode, Args: preparatorArgs}
+	return &[]Preparator{formatCodePreparator}
+}
+
+// formatCode formats go code
+func formatCode(args ...interface{}) error {
+	filePath := args[0].(string)
+	file := filepath.Base(filePath)
```
Comment (daria-malkova) [inline playground/backend/internal/preparators/go_preparators.go:35]: if methods return no errors you could directly use` cmd.Dir = filepath.Dir(filePath)`

File: playground/backend/internal/preparators/go_preparators_test.go [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+	err := setupPreparedFiles()
+	if err != nil {
+		logger.Fatal(err)
+	}
+	defer teardown()
+	m.Run()
+}
+
+func setupPreparedFiles() error {
+	correctCode := "package main\n\nimport \"fmt\"\n\nfunc main() {\n\tfmt.Println(\"hello world\")\n\n\n}\n"
```
Comment (daria-malkova) [inline playground/backend/internal/preparators/go_preparators_test.go:44]: constant

File: playground/backend/internal/preparators/go_preparators_test.go [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+	err := createFile(correctFile, correctCode)
+	if err != nil {
+		return err
+	}
+	incorrectCode := "package main\n\nimport \"fmt\"\n\nfunc main() {\n\tfmt.Println(\"hello world\"\n\n}\n"
```
Comment (daria-malkova) [inline playground/backend/internal/preparators/go_preparators_test.go:49]: constant

File: playground/backend/internal/preparators/go_preparators_test.go [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+		t.Run(tt.name, func(t *testing.T) {
+			if got := GetGoPreparators(tt.args.filePath); !reflect.DeepEqual(fmt.Sprint(got), fmt.Sprint(tt.want)) {
+				t.Errorf("GetGoPreparators() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+func Test_formatCode(t *testing.T) {
+	preparedArgs1 := make([]interface{}, 1)
```
Comment (daria-malkova) [inline playground/backend/internal/preparators/go_preparators_test.go:109]: зlease extract duplicate code to small help function

**Review by daria-malkova — CHANGES_REQUESTED** [review]

File: playground/backend/internal/preparators/go_preparators_test.go [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+			args:    args{preparedArgs2},
+			wantErr: true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if err := formatCode(tt.args.args...); (err != nil) != tt.wantErr {
+				t.Errorf("formatCode() error = %v, wantErr %v", err, tt.wantErr)
+			}
+		})
```
Comment (KhaninArtur) [inline playground/backend/internal/preparators/go_preparators_test.go:146]: Can we also check that the code was formatted as we expect?

**Review by KhaninArtur — CHANGES_REQUESTED** [review]

File: playground/backend/internal/preparators/go_preparators_test.go (hunk 8 above)
Comment (ilya-kozyrev) [reply playground/backend/internal/preparators/go_preparators_test.go]: for tests, it's not significant

File: playground/backend/internal/preparators/go_preparators_test.go (hunk 11 above)
Comment (pavel-avilov) [reply playground/backend/internal/preparators/go_preparators_test.go:146]: In theory, yes, but you need to parse the output and draw conclusions from it. And I don’t think that there are no cases when fmt will format it as something wrong.

File: playground/backend/internal/preparators/go_preparators_test.go (hunk 5 above)
Comment (pavel-avilov) [reply playground/backend/internal/preparators/go_preparators_test.go:136]: Done.

File: playground/backend/internal/preparators/go_preparators_test.go (hunk 2 above)
Comment (pavel-avilov) [reply playground/backend/internal/preparators/go_preparators_test.go:45]: Done.

File: playground/backend/internal/preparators/go_preparators_test.go (hunk 10 above)
Comment (pavel-avilov) [reply playground/backend/internal/preparators/go_preparators_test.go]: Done

File: playground/backend/internal/preparators/go_preparators_test.go (hunk 9 above)
Comment (pavel-avilov) [reply playground/backend/internal/preparators/go_preparators_test.go]: Done

File: playground/backend/internal/preparators/go_preparators_test.go (hunk 8 above)
Comment (pavel-avilov) [reply playground/backend/internal/preparators/go_preparators_test.go]: Done

File: playground/backend/internal/preparators/go_preparators.go (hunk 6 above)
Comment (pavel-avilov) [reply playground/backend/internal/preparators/go_preparators.go]: Done

File: playground/backend/internal/preparators/go_preparators_test.go (hunk 3 above)
Comment (pavel-avilov) [reply playground/backend/internal/preparators/go_preparators_test.go]: Done

File: playground/backend/internal/preparators/go_preparators_test.go (hunk 4 above)
Comment (pavel-avilov) [reply playground/backend/internal/preparators/go_preparators_test.go]: Done

File: playground/backend/internal/preparators/go_preparators_test.go (hunk 11 above)
Comment (KhaninArtur) [reply playground/backend/internal/preparators/go_preparators_test.go:146]: Fair enough, we should trust `go fmt`🙂

**Pull request closed without merging**